Device roles to capabilities (#197)
margo/specification: system-design/specification/margo-management-interface/device-capabilities.md

| vendor        | string    | Y    | Defines the device vendor.|
| modelNumber        | string    | Y    | Defines the model number of the device.|
| serialNumber       | string    | Y    | Defines the serial number of the device.|

> | cpu | []CPU | N* | List of CPU entries available on the device. Utilized to match with the required resources defined in the application description. See the [CPU](#cpu-attributes) section below.|

Reviewer: I think I would change all `N*` to `Y*` since no is the exception to the rule instead of the rule.

Author: addressed in the latest commit. Good suggestion.

> | cpu | []CPU | Y* | List of CPU entries available on the device. Utilized to match with the required resources defined in the application description. See the [CPU](#cpu-attributes) section below.|

Reviewer 2: Now that `cpu` supports an array, we should probably rename this to `cpus`. The change is breaking anyhow...

| memory | string | Y* | The amount of memory available for applications to utilize on the device. The value is given in binary units (`Ki` = Kibibytes, `Mi` = Mebibytes, `Gi` = Gibibytes). This is defined by the device owner.|
| storage | string | Y* | The amount of storage available for applications to utilize on the device. The value is given in binary units (`Ki` = Kibibytes, `Mi` = Mebibytes, `Gi` = Gibibytes, `Ti` = Tebibytes, `Pi` = Pebibytes, `Ei` = Exbibytes). This is defined by the device owner.|
| peripherals | []Peripheral | Y* | Peripherals available for applications to utilize on the device. See the [Peripheral](#peripheral-attributes) section below.|
| interfaces | []CommunicationInterface | Y* | Communication interfaces available for applications to utilize on the device. See the [Communication Interfaces](#communicationinterface-attributes) section below.|

> | otelCollector | boolean | Y* | Reports whether an OpenTelemetry (OTEL) collector is present on the device. |

Reviewer 2: ```suggestion
| otelCollector | boolean | Y* | Reports whether an OpenTelemetry (OTEL) collector is present on the device. A device that is capable of hosting workloads MUST report `true`. |
```

> | supportedRuntimes | []SupportedRuntime | Y* | Supported workload runtimes present on the device. See the [SupportedRuntime](#supportedruntime) definition for all permissible values. A device that hosts workloads MUST report at least one entry.|

Reviewer 2: ```suggestion
| supportedRuntimes | []SupportedRuntime | Y* | Supported workload runtimes present on the device. See the [SupportedRuntime](#supportedruntime) definition for all permissible values. A device that is capable of hosting workloads MUST report at least one entry.|
```

> | supportedDeploymentTypes | []SupportedDeploymentType | Y* | The deployment profile types the device can receive and process locally. See the [SupportedDeploymentType](#supporteddeploymenttype) definition for all permissible values. A device that hosts workloads MUST report at least one entry.|

Reviewer 2: ```suggestion
| supportedDeploymentTypes | []SupportedDeploymentType | Y* | The deployment profile types the device can receive and process locally. See the [SupportedDeploymentType](#supporteddeploymenttype) definition for all permissible values. A device that is capable of hosting workloads MUST report at least one entry.|
```

> | supportedDeploymentTypes | []SupportedDeploymentType | N* | The manifest/deployment formats the device can receive and process locally. See the [SupportedDeploymentType](#supporteddeploymenttype) definition for all permissible values. A device that hosts workloads MUST report at least one entry.|

Reviewer: Maybe even link "deployment profile types" to the application description page

```suggestion
| supportedDeploymentTypes | []SupportedDeploymentType | N* | The deployment profile types the device can receive and process locally. See the [SupportedDeploymentType](#supporteddeploymenttype) definition for all permissible values. A device that hosts workloads MUST report at least one entry.|
```

> > Note:  \* A see-through gateway not hosting workloads itself MUST omit or provide empty values for these fields. The WFM infers such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy. 

Reviewer 2: ```suggestion
> Note:  \* A see-thru gateway not hosting workloads itself MUST omit or provide empty values for these fields. The WFM infers such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy. 
```

Reviewer 2: Shouldn't we decide upon omit vs empty values in the spec?

> ### CPU Attributes

Reviewer 2: With `cpu` / `cpus` now accepting an array, it's probably worth spending a sentence to describe how WFMs are expected to match that array with a single-valued `cpu` from an app description. Maybe add something like this:

> Each entry in `cpus` describes one set of cores that a single workload can be placed on. A device satisfies an application's `requiredResources.cpu` only if one entry provides at least the required number of `cores` and, when the application specifies architectures, that entry's `architecture` is one of the listed values. Cores from separate entries MUST NOT be added together when evaluating the requirement.

> > Note:  \* Required for devices that host workloads locally. A device that does not host workloads itself, for example see-thru gateway, that only relays the devices behind it — omits these fields. The WFM infers that such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy. No explicit gateway flag is required.

Reviewer: This change in combination with changing `N*` to `Y*`

```suggestion
> Note:  \* A see-through gateway not hosting workloads itself MUST omit or provide empty values for these fields. The WFM infers such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy. 
```

CPU element defining the device's CPU characteristics.

| Attribute | Type | Required? | Description |
| --- | --- | --- | --- |

> | cores | integer |  Y  | Defines the cores available within the hosts CPU. Specified as decimal units of CPU cores (e.g., `0.5` is half a core). This is defined by the device owner. After deployment of the application, the device MUST provide this number of CPU cores for the application.|

Reviewer 2: 0.5 wouldn't be an accepted value for `cores` under type `integer`. See my note in the OpenAPI spec below.

| peripherals | []Peripheral | Y* | Peripherals available for applications to utilize on the device. See the [Peripheral](#peripheral-attributes) section below.|
| interfaces | []CommunicationInterface | Y* | Communication interfaces available for applications to utilize on the device. See the [Communication Interfaces](#communicationinterface-attributes) section below.|
| otelCollector | boolean | Y* | Reports whether an OpenTelemetry (OTEL) collector is present on the device. A device that is capable of hosting workloads MUST report `true`. |
| supportedRuntimes | []SupportedRuntime | Y* | Supported workload runtimes present on the device. See the [SupportedRuntime](#supportedruntime) definition for all permissible values. A device that is capable of hosting workloads MUST report at least one entry.|
| supportedDeploymentTypes | []SupportedDeploymentType | Y* | The deployment profile types the device can receive and process locally. See the [SupportedDeploymentType](#supporteddeploymenttype) definition for all permissible values. A device that is capable of hosting workloads MUST report at least one entry.|

> > Note:  \* A see-thru gateway not hosting workloads itself MUST omit these fields or provide empty values for them, reporting `otelCollector` as `false`. The WFM infers such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy.

Reviewer 2: Sending empty values for `supportedRuntimes` and `supportedDeploymentTypes` is not allowed per OpenAPI spec, so I'd settle on omit.

```suggestion
> Note:  \* A see-thru gateway not hosting workloads itself MUST omit these fields. The WFM infers such a device is non-hosting from the absence of these capabilities, and infers a gateway relationship from the parent/child `deviceId` hierarchy.
```

| microphone | This type stands for a microphone peripheral. |
| speaker | This type stands for a speaker peripheral. |

#### SupportedRuntime

| Permissible Values | Description |
| --- | --- |
| oci | OCI container runtime. |

#### SupportedDeploymentType

| Permissible Values | Description |
| --- | --- |
| helm | Device can receive and process Kubernetes Helm chart deployments locally. |

> | compose | Device can receive and process Docker Compose service deployments locally. |

Reviewer: ```suggestion
| compose | Device can receive and process Compose service deployments locally. |
```

### Opaque gateways

A device may represent, and aggregate the capabilities of, multiple child-devices behind it and report itself as a single Margo device to the WFM. This type of device is referred to as an opaque gateway. Opaque gateways report the combined capabilities of all the devices they connect to the WFM as a single `DeviceCapabilitiesManifest`. Because the child-devices are not individually visible to the WFM, an opaque gateway is seen as a single device and reports the aggregated resource fields, `supportedRuntimes`, and `supportedDeploymentTypes` of the devices behind it.

> Example: An opaque gateway has two child-devices. Each child-device has an ARM64 processor with 2 cores, 5 GB of memory, 32 GB of storage, and 1 ethernet interface. The gateway will report capabilities of 2 CPUs (arm64) with 2 cores each, 10 GB of memory, 64 GB of storage, and 2 ethernet interfaces. Since the gateway can deploy compose applications on its child-devices it will report `supportedDeploymentTypes: ["compose"]`.

### See-thru gateways

> WFM clients may connect one or more child-devices to the WFM while allowing the WFM to see each device behind it as an individual device with its own capabilities. This type of client is referred to as a **see-thru gateway**. A see-thru gateway is not a special device type; it is treated as an ordinary device that additionally reports the capabilities of the devices behind it. The WFM infers the gateway relationship from the reported information — the parent/child `deviceId` hierarchy and it is especially evident when the gateway itself reports no workload-hosting capabilities.

Reviewer 2: > A see-thru gateway is not a special device type

That's true from a payload perspective. From a conformance perspective, however, the rules for see-thru gateways are relaxed (per device requirements above).

> it is especially evident when the gateway itself reports no workload-hosting capabilities

Does the spec *forbid* see-thru gateways to report runtimes if they can host workloads, or are those gateways just not *required* to report runtimes?

A see-thru gateway MUST report its own capabilities and the capabilities of each device it connects to the WFM. This is done by calling the `device capabilities` endpoint for the gateway itself and for each device behind the gateway. The `deviceId` in the endpoint is used to indicate the hierarchy of devices, with a parent/child relationship. For example, if a see-thru gateway with `deviceId` "gateway1" connects two devices with `deviceId` "deviceA" and "deviceB", the gateway would call the `device capabilities` endpoint three times with the following `deviceId`s: "gateway1", "gateway1/deviceA", and "gateway1/deviceB".

> A see-thru gateway that does not host workloads itself omits the workload-hosting fields (`cpu`, `memory`, `storage`, `otelCollector`, `supportedRuntimes`, `supportedDeploymentTypes`), so its manifest contains only the required identity fields. A see-thru gateway that is also capable of hosting workloads reports the workload-hosting fields, including at least one entry in both `supportedRuntimes` and `supportedDeploymentTypes`.

Reviewer: ```suggestion
A see-thru gateway that does not host workloads itself MUST omit the workload-hosting fields (`cpu`, `memory`, `storage`, `otelCollector`, `supportedRuntimes`, `supportedDeploymentTypes`), so its manifest contains only the required identity fields. A see-thru gateway that is also capable of hosting workloads reports the workload-hosting fields, including at least one entry in both `supportedRuntimes` and `supportedDeploymentTypes`.
```

> A see-thru gateway that does not host workloads itself MUST omit the workload-hosting fields (`cpu`, `memory`, `storage`, `otelCollector`, `supportedRuntimes`, `supportedDeploymentTypes`), so its manifest contains only the required identity fields. A see-thru gateway that is also capable of hosting workloads reports the workload-hosting fields, including at least one entry in both `supportedRuntimes` and `supportedDeploymentTypes`.

Reviewer 2: See above. This is a stricter and different requirement than what the note tells above.

> Example: An opaque gateway has two child-devices. Each child-device has an ARM64 processor with 2 cores, 5 GB of memory, 32 GB of storage, and 1 ethernet interface. The gateway will report capabilities of 2 CPUs (arm64) with 2 cores each, 10 GB of memory, 64 GB of storage, and 2 ethernet interfaces. Since the gateway can deploy compose applications on its child-devices it will report `supportedDeploymentTypes: ["compose"]`.

### See-thru gateways

WFM clients may connect one or more child-devices to the WFM while allowing the WFM to see each device behind it as an individual device with its own capabilities. This type of client is referred to as a **see-thru gateway**.

> A see-thru gateway uses the same `DeviceCapabilitiesManifest` schema as any other device — from a payload perspective it is an ordinary device that also reports the devices behind it. Its conformance rules are relaxed, though: unlike a standalone hosting device, a see-thru gateway is not required to host workloads and need not report workload-hosting capabilities. The WFM infers the gateway relationship from the parent/child `deviceId` hierarchy, which is typically most evident when the gateway reports no workload-hosting capabilities.

Reviewer 2: ```suggestion
A see-thru gateway uses the same `DeviceCapabilitiesManifest` schema as any other device — from a payload perspective it is an ordinary device that also reports the devices behind it. Its conformance rules are relaxed, though: unlike non-gateway device, a see-thru gateway is not required to host workloads and need not report workload-hosting capabilities. The WFM infers the gateway relationship from the parent/child `deviceId` hierarchy, which is typically most evident when the gateway reports no workload-hosting capabilities.
```

**How a see-thru gateway reports capabilities**

A see-thru gateway MUST report its own capabilities and the capabilities of each device it connects to the WFM:

1. Call the `device capabilities` endpoint once for the gateway itself, then once for each device behind it.
2. Encode the hierarchy in the `deviceId` as a parent/child path. For example, a gateway `gateway1` with two child-devices calls the endpoint three times, with `deviceId`s `gateway1`, `gateway1/deviceA`, and `gateway1/deviceB`.
3. Report the gateway's own manifest **before** any child manifest. If the WFM receives a child manifest first, it MUST reject the request with a `404 Not Found` response code.

**What the gateway reports about itself**

| If the gateway... | Then its own manifest MUST... |
| --- | --- |

> | does **not** host workloads | contain only the required identity fields — omit or provide empty values for the workload-hosting fields (`cpus`, `memory`, `storage`, `peripherals`, `interfaces`, `supportedRuntimes`, `supportedDeploymentTypes`), and omit `otelCollector` or report it as `false` |

Reviewer 2: See above. Regarding otelCollector: since the observability part of the Margo spec does not define what a collector on a non-hosting device collects, reporting `true` for otelCollector on a non-hosting device would have no contract.

```suggestion
| does **not** host workloads | contain only the required identity fields — omit the workload-hosting fields (`cpus`, `memory`, `storage`, `peripherals`, `interfaces`, `supportedRuntimes`, `supportedDeploymentTypes`), and omit `otelCollector` |
```
